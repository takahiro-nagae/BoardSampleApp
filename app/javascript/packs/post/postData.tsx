/****************************************
 * 投稿の定義クラス
 ****************************************/
export class PostData {
  /** ポストID */
  id: number;
  /** カテゴリID */
  category_id: string;
  /** 非表示フラグ */
  hide_flag: string;
  /** IPアドレス */
  ip: string;
  /** 名前 */
  name: string;
  /** 名前 */
  mail: string;
  /** 件名 */
  subject: string;
  /** 本文 */
  text: string;
  /** 投稿日時 */
  created_at: Date;
  /** クライアントIPアドレス */
  clientIp: string;
}