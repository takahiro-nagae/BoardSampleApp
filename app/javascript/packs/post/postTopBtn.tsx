import * as React from 'react';
import { useHistory } from 'react-router-dom';

/****************************************
 * 投稿一覧画面上部のボタンコンポーネント
 ****************************************/
export const PostTopBtn = () => {
    const history = useHistory();

    // ================================
    // 戻るイベント
    // ================================
    const back = () => {
      history.push('/board/index');
    };

    return (
      <button type='button' className='btn btn-secondary' onClick={back}>
        戻る
      </button>
    );
};