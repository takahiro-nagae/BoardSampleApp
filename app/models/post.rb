class Post < ApplicationRecord
    # ============================
    # validate:uuid
    # ============================
    # 必須項目
    validates :uuid, presence: { message: "投稿処理が失敗しました。ページを更新して再度お試しください。" }
    # ============================
    # validate:name
    # ============================
    # 最大桁数:256
    validates :name, length: {maximum: 256, message: "名前は256文字以内で入力してください"}
    # ============================
    # validate:mail
    # ============================
    # 最大桁数:256
    validates :mail, length: {maximum: 256, message: "メールアドレスは256文字以内で入力してください"}
    # ============================
    # validate:subject
    # ============================
    #　最大桁数:256
    validates :subject, length: {maximum: 256, message: "件名は256文字以内で入力してください"}
    # ============================
    # validate:text
    # ============================
    # 必須項目
    validates :text, presence: { message: "本文は必須です" }
    # 最大桁数:10000
    validates :text, length: {maximum:10000, message: "本文は10000文字以内で入力してください"}
end
