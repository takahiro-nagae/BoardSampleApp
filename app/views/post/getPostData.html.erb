<h1>Post#get</h1>
<p>Find me in app/views/post/get.html.erb</p>
