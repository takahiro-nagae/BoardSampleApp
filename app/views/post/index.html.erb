<h1 id="title" class="my-5 display-4 text-primary">掲示板</h1>
<div id="app"></div>
<%= javascript_pack_tag 'index' %>