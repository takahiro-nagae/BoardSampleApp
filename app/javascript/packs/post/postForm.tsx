import * as React from 'react';
import { useForm } from 'react-hook-form'
import { csrfToken } from 'rails-ujs';
import axios from 'axios'
import { useState, useEffect, Fragment} from 'react';
import { PostData } from './postData';

/****************************************
 * 新規投稿コンポーネント
 ****************************************/
export const PostForm = (props) => {
    const {category_id, updatePosts, loginUser} = props;
    /** 名前 */
    const [name, setName] = useState('');
    /** メール */
    const [mail, setEmail] = useState('');

    // ================================
    // バリデーション設定
    // ================================
    const {
        register,
        handleSubmit,
        formState: { errors }
      } = useForm<PostData>({
        mode: "onChange",
        criteriaMode: "all",
        shouldFocusError: false,
    });

    // ================================
    // form送信イベント
    // ================================
    const onSubmit = () => {
        axios.defaults.headers.common['X-CSRF-Token'] = csrfToken();
        axios.post('http://localhost:3000/post/regist', new FormData(document.forms['postForm']))
        .then(res => {
            // postList更新
            updatePosts(res.data.post);
            // 初期化
            document.getElementById('error_message').innerHTML='';
            document.forms['postForm'].reset();
        })
        .catch(error => {
            // サーバサイドバリデーションをレンダリング
            document.getElementById('error_message').innerHTML=error.response.data.errors.text;
        });
    }

    useEffect(() => {
        if(loginUser!= undefined) {
            setName(loginUser.name);
            setEmail(loginUser.email);
        }
    }, [loginUser]);

    return(
        <>
            <h2 className='bg-primary text-white p-1'>新規投稿</h2>
            <div id="error_message" className='text-danger my-4'></div>
            <form onSubmit={handleSubmit(onSubmit)} id='postForm' className='mb-5'>
                <input type='hidden' name='category_id' value={category_id} />
                <div className='form-group'>
                    <label htmlFor='name'>名前</label>
                    <input type='text' className='form-control' id='name' name='name' placeholder='名無し'
                         {...register("name", { maxLength: 256 })} value={name}
                         onChange={event => setName(event.target.value)} />
                    {errors.name?.types?.maxLength && <span className='text-danger'>名前は256文字以内で入力してください</span>}
                </div>
                <div className='form-group'>
                    <label htmlFor='mail'>メールアドレス</label>
                    <input type='email' className='form-control' id='mail' name='mail' placeholder='aaa@mail.com'
                     {...register("mail", { maxLength: 256 })} value={mail} onChange={event => setEmail(event.target.value)} />
                    {errors.mail?.types?.maxLength && <span className='text-danger'>メールアドレスは256文字以内で入力してください</span>}
                </div>
                <div className='form-group'>
                    <label htmlFor='subject'>件名</label>
                    <input type='text' className='form-control' id='subject' name='subject'
                         placeholder='〇〇について' {...register("subject", { maxLength: 256 })} />
                    {errors.subject?.types?.maxLength && <span className='text-danger'>件名は256文字以内で入力してください</span>}
                </div>
                <div className='form-group'>
                    <label htmlFor='subject'>件名</label>
                    <textarea className='form-control'
                        id='text' name='text' {...register("text", { required: true, maxLength: 10000 })} />
                    {errors.text?.types?.required && <span className='text-danger'>本文は必須です</span>}
                    {errors.text?.types?.maxLength && <span className='text-danger'>本文は10000文字以内で入力してください</span>}
                </div>
                <button type='submit' className='btn btn-success'>投稿</button>
            </form>
        </>
    );
};