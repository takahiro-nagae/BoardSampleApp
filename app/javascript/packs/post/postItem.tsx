import * as React from 'react';
import { Fragment } from 'react';
import { Link } from 'react-router-dom';
import { csrfToken } from 'rails-ujs';
import axios from 'axios'
import { PostData } from './postData'
import '../../../assets/stylesheets/post.scss'

/****************************************
 * 投稿一覧の各行データコンポーネント
 ****************************************/
export const PostItem = (props: PostData) => {

  const {post_id, category_id, hide_flag, ip, name, subject, text, post_date, clientIp, updatePosts, loggedInStatus} = props;

  // ================================
  // 投稿削除（非表示）イベント
  // ================================
  const deletePost = () => {

    let fd = new FormData();
    fd.append('category_id', category_id);
    fd.append('post_id', post_id.toString());

    axios.defaults.headers.common['X-CSRF-Token'] = csrfToken();
    axios.post('http://localhost:3000/post/delete', fd)
    .then(res => {
        // postList更新
        updatePosts(res.data);
        // formを初期化
        document.forms['postForm'].reset();
    })
    .catch(error => console.log(error))
  }

  return(
    <>
      { (hide_flag == '0' || loggedInStatus) &&
        <li className='list-group-item'>
          <p className='text-muted'>
            <u>
              <span>{post_id} :{name}　{subject}　{post_date}</span>
              {
                (hide_flag == '1' && loggedInStatus) &&
                <span className='text-danger'>　削除済み</span>
              }
            </u>
          </p>
          <div className='d-flex justify-content-between align-items-start'>
            <p className='white-space'>{text}</p>
            { (ip == clientIp || loggedInStatus) &&
              <>
                <div className='text-right'>
                <button type='button' onClick={deletePost} className={'btn btn-danger' + ' delete' + post_id}>削除</button>
                </div>
              </>
            }
          </div>
        </li>
      }
    </>

  );
}