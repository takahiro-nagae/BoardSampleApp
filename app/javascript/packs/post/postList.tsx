import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { useEffect, useState, Fragment } from 'react';
import { useLocation } from 'react-router-dom';
import queryString from 'query-string';
import axios from 'axios';

import { PostData } from './postData'
import { PostItem } from './postItem'
import { PostTopBtn } from './postTopBtn';
import { PostForm } from './postForm';
import { Error } from '../error/error';

/****************************************
 * 投稿一覧コンポーネント
 ****************************************/
export const PostList = (props) => {
    // props
    const {loggedInStatus, loginUser, setTitle} = props;
    //投稿のstate
    const [posts, setPosts] = useState<PostData[]>([]);
    const [clientIp, setClientIp] = useState('');

    const query = queryString.parse(useLocation().search);

    useEffect(() => {
        // タイトルをカテゴリ名に変更
        setTitle(query['category_name']);
        // クライアントIP取得
        axios.get('https://geolocation-db.com/json/').then((res) => {
            setClientIp(res.data.IPv4);
        });
        // 投稿一覧を取得
        getPostList();
    }, []);

    // ログイン状態が変化したら投稿一覧を再取得
    useEffect(() => {
        getPostList();
    }, [loggedInStatus]);

    const getPostList = () => {
        // 投稿一覧を取得
        axios.get<PostData[]>('http://localhost:3000/post/get?category_id=' + query['category_id'])
        .then((res) => {
            setPosts(res.data);
        })
        .catch(error => {
            if(error.response.status == 422) {
                ReactDOM.render(
                    <Error msg={error.response.data.errors} />
                    , document.getElementById('postContent')
                );
            }
        })
    }


    // ================================
    // 子コンポーネントから投稿を再設定
    // ================================
    const updatePosts = (posts: PostData[]) => {
        let newPosts = [...posts];
        setPosts(newPosts);
    }
    return(
        <>
            {/* 戻るボタン */}
            <PostTopBtn />
            <div id='postContent'>
                {/* 投稿一覧 */}
                <div className='py-5'>
                    <ul className='list-group'>
                        {posts.map(post => (
                            <PostItem postData={post} clientIp={clientIp} updatePosts={updatePosts} key={post.id} />
                        ))}
                    </ul>
                </div>
                {/* 投稿フォーム */}
                <PostForm category_id={query['category_id']} updatePosts={updatePosts}
                    clientIp={clientIp} loginUser={loginUser} />
            </div>
        </>
    );
};