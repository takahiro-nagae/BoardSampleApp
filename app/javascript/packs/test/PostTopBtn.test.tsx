import * as renderer from 'react-test-renderer';
import { PostTopBtn } from '../post/postTopBtn';

test('投稿一覧トップボタン：コンポーネントテスト', () => {
    const component = renderer.create(<PostTopBtn />);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});