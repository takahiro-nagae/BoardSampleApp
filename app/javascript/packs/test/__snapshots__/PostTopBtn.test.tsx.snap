// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`投稿一覧トップボタンのコンポーネントテスト 1`] = `
<button
  className="btn btn-secondary"
  onClick={[Function]}
  type="button"
>
  戻る
</button>
`;

exports[`投稿一覧トップボタン：コンポーネントテスト 1`] = `
<button
  className="btn btn-secondary"
  onClick={[Function]}
  type="button"
>
  戻る
</button>
`;
